### ISSUE TITLE
Integrate darts models with MLflow (logging, autologging, loading)

### ISSUE THREAD

**Issue by pelekhs** [unit8co/darts#1618]
**Is your feature request related to a current problem? Please describe.**
Missing API to log / load / autolog ML models to MLflow.

**Describe proposed solution**
Provide an API that integrates darts models with MLflow models and provides model logging and loading capabilities. (check this[](https://mlflow.org/docs/latest/python_api/mlflow.pytorch.html) as an example for pytorch)

**Describe potential alternatives**
Pyfunc models and model flavors can be used right now but this quite a time consuming process to handle darts models with MLflow. 

**Additional context**
Would it be maybe good to handle darts models as pytorch lightning models during this process? And maybe build up on the prexisting pytorch integration of MLflow?

**Comment by dennisbader** [comment]
@pelekhs, acknowledged. Give us some time to go over this and come back to you.

**Comment by turbotimon** [comment]
I would really appreciate that too. I am currently using MLFlow together with Darts for a project, but have to do it manually

**Comment by dennisbader** [comment]
Hi again. We discussed and think this would be a great idea :)
Can you help us understand what such an integration would take?

In general we are open to have something like this in place as long as it doesn't disrupt our current API. Some compromises might be fine though if they make sense.

From #5689 I see that you're already in contact with the MLFlow team. We would be happy to offer guidance/supervision as well!

As an extension to the feature request description in #5689: 
- All of Darts TorchForecastingModels (TFM, deep learning) are built on top of PyTorch Lightning (and underlying PyTorch). So callbacks such as MLFlow loggers are available to the user.
- In Darts, the TFM models perform the PyTorch-Lightning-related setup (trainer, datasets, etc) behind the hood. This means that our API is different to the standard PyTorch-Lightning API.

**Comment by turbotimon** [comment]
In my opinion, the main advantage is to get the [mlflow.autolog()](https://mlflow.org/docs/latest/python_api/mlflow.html#mlflow.autolog) functionality working
- As it already works with [Pytorch Lightning](https://mlflow.org/docs/latest/tracking.html?highlight=lightning#pytorch), it maybe needs just a thin wrapper
- When I run it with e.g. TCNModel, I get the following error (I didn't investigate further):
  `WARNING mlflow.utils.autologging_utils: Encountered unexpected error during pytorch autologging: 'LightningDoublePrecisionModule' object has no attribute 'example_input_array'`.
- Customizations I can think of:
  - Log (save/export) the full darts model, not only the underlying Pytorch model.
  - Logging some darts-specific or useful metrics and parameters.

**Comment by Sonta** [comment]
Same here, would appreciate this.
@turbotimon if it helps you, I had the exact same error until i converted my time series to 32-bit.

Getting another error about a `NoneType` instead, but the artifact logging seems to work nevertheless.

My main issue is the loading of the model which I would need to implement manually through a custom mlflow pyfunc (haven't done it because I have a workaround through another model registry).

Can offer some assistance if someone wants to take the lead on this, don't think I'll have the capacity to do it by myself though. Looking at the responses from the mlflow guys it would need to be some kind of 3rd-party package I guess.

**Comment by turbotimon** [comment]
@Sonta Thanks for the hint with 32-bit! I'll try it out

**Comment by chartsengrafs** [comment]
We need mlflow for Darts and time series so badly.

**Comment by GitHunter0** [comment]
> We need mlflow for Darts and time series in general so badly.

Yes, this is essential to enable `darts` to be used in a production setting.

**Comment by GitHunter0** [comment]
In the meanwhile, could some please provide an example converting a `darts` model into an `MLflow` model using `mlflow.pyfunc.PythonModel`?

**Comment by d-sooter** [comment]
Hi there we are looking to set up a solution using mlflow and darts. Does anyone here have any tips or sample projects. Or a general structure we could follow.

We would also be very interested in this intergration and would be willing to help make it happen and/or test it.

**Comment by turbotimon** [comment]
> Hi there we are looking to set up a solution using mlflow and darts. Does anyone here have any tips or sample projects. Or a general structure we could follow.
> 
> We would also be very interested in this intergration and would be willing to help make it happen and/or test it.

As mlflow is not integrated in Darts, the convenient [autolog](https://mlflow.org/docs/latest/tracking.html#automatic-logging) will not work, so logging has to be done manually (with `log_param`. `log_metric`, `log_artifact` and so one). But that's no too complicated either, just take the [official example](https://mlflow.org/docs/latest/tutorials-and-examples/tutorial.html#training-the-model) from mlflow and substitute the sklearn part with Darts. Model registry will probably also not work, but i posted a workaround for that above.

So a simplified structure would be:
```python
# 0. setup
import mlflow
import os

os.environ["MLFLOW_TRACKING_USERNAME"] = "<user>"
os.environ["MLFLOW_TRACKING_PASSWORD"] = "<pw>"
mlflow.set_tracking_uri("") # local, else set url here
mlflow.set_experiment("My_Experiment")

# 1. start run
RUN_NAME = "MyRun"
mlflow.start_run(run_name=RUN_NAME)

# 2. fit&evaulate a model
....

# 3. log your stuff
mlflow.log_param("epochs", 15)
mlflow.log_metric("accuracy", 0.99)
mlflow.log_artifact("mymodel.pickle") # if saved e.g. with model.save("mymodel.pickle")
...

# 4. end
mlflow.end_run() 
```

Does this help or where do you need help in particular?

**Comment by d-sooter** [comment]
Thanks a lot for the answer. Still getting going with both. I was also looking into storing the model with mlflow. While i wont be using mlflow to serve the model, i like their way of storing it. I still haven’t looked at storing and retrieving a darts model with mlflow. But since they both use picke i cant Imagine it would be an issue.

**Comment by Houcemderbel** [comment]
@pelekhs do you have an example of reading and writing models to mlflow? we are currently trying to set up tracking in mlflow but would like to continue to use Darts. 

We want to use external storage to store the artifacts (s3/azure blob) and these are the general steps im thinking but if you have an example it would be very helpful.

Training step:
- create pickle model using darts. (does it have to be created as a local file or can i create it on the storage directly)
- log artifact ex: mlflow.log_artifact("mymodel.pkl")

Prediction step:
- load model 
- build darts model
- run prediction

if you have any sample code it would be awesome. 
PS im new to mlflow

**Comment by thuyng-ing** [comment]
Hi, is there any update on this integration? Thank you very much!

**Comment by cargecla1** [comment]
Hello everyone,

I have a solution to the MLflow integration via additional code on you notebook/ .py file making pytorch calls and activating tensorboard collection on Darts' s pytortch models. Could anyone from the team reach me out to let me know where I can add an MLflow notebook example that solves this issue? I have examples with NBEATS, TCN, NHiTS, DLinear, and NLinear doing autologging with MLflow code added. Please reach out.

**Comment by madtoinou** [comment]
Hi @cargecla1,

If you think that there is enough content to make a whole notebook, you can create a new one in the `examples/` folder.

If the code is rather generic and similar across the models, it would probably be better to just add a section in the example [user guide](https://unit8co.github.io/darts/userguide/torch_forecasting_models.html) (source is in `docs/userguide/`) or directly in the notebook of the torch-based models (RNN, TCN, Transformer, NBeats, Tide, ...).

**Comment by cargecla1** [comment]
> Hi @cargecla1,
> 
> If you think that there is enough content to make a whole notebook, you can create a new one in the `examples/` folder.
> 
> If the code is rather generic and similar across the models, it would probably be better to just add a section in the example [user guide](https://unit8co.github.io/darts/userguide/torch_forecasting_models.html) (source is in `docs/userguide/`) or directly in the notebook of the torch-based models (RNN, TCN, Transformer, NBeats, Tide, ...).
> 
> It would be great if you could open a draft PR as soon as possible so that we can start collecting feedback on the integration with MLflow :)

I will create a draft PR over the weekend, thank you for your feedback!

**Comment by cargecla1** [comment]
> Hi @cargecla1,
> 
> If you think that there is enough content to make a whole notebook, you can create a new one in the `examples/` folder.
> 
> If the code is rather generic and similar across the models, it would probably be better to just add a section in the example [user guide](https://unit8co.github.io/darts/userguide/torch_forecasting_models.html) (source is in `docs/userguide/`) or directly in the notebook of the torch-based models (RNN, TCN, Transformer, NBeats, Tide, ...).
> 
> It would be great if you could open a draft PR as soon as possible so that we can start collecting feedback on the integration with MLflow :)

Hello,

I added Preview PR here: [cargecla1:feature/MLflow_1618](https://github.com/cargecla1/darts/tree/feature/MLflow_1618).

Let me know what you think.

**Comment by dennisbader** [comment]
Hi @cargecla1, and thanks for adding the example. I created Draft PR #2092 for this.

**Comment by cargecla1** [comment]
> Hi @cargecla1, and thanks for adding the example. I created Draft PR #2092 for this.

Hello @dennisbader, excellent, thank you. Glad to be able to support.

**Comment by dnerini** [comment]
Hi there, I looked into #2092 but I don't understand how the DARTS model can be loaded from MLflow once logged as a pytorch model, am I missing on something obvious here? Thanks for the help!

**Comment by cargecla1** [comment]
> Hi there, I looked into #2092 but I don't understand how the DARTS model can be loaded from MLflow once logged as a pytorch model, am I missing on something obvious here? Thanks for the help!

Hello @dnerini 

Please refer to this link to see how to load the model after you register it: https://mlflow.org/docs/latest/model-registry.html

This page shows you how to use the MLflow UI to manage your registered models, there are two ways to log a model: using the mlflow.<model_flavor>.log_model() method or using the mlflow.register_model() method, above link explains this.

Please pay special attention to the subtitle: 
[Fetching an MLflow Model from the Model Registry](https://mlflow.org/docs/latest/model-registry.html#id11)

and

[Serving an MLflow Model from Model Registry](https://mlflow.org/docs/latest/model-registry.html#id12)

Hope this helps!

Cheers

**Comment by turbotimon** [comment]
> > Hi there, I looked into #2092 but I don't understand how the DARTS model can be loaded from MLflow once logged as a pytorch model, am I missing on something obvious here? Thanks for the help!
> 
> Hello @dnerini
> 
> Please refer to this link to see how to load the model after you register it: https://mlflow.org/docs/latest/model-registry.html
> 
> ...

I don't think that solve the issue of @dnerini. Because in the current state of #2092 (13d8113) there is no call to `log_model()`, which saves the model as an artifact, there is no sense in `register_model` as the given `model_uri` doesn not point to a valid model artifact. I tried it out and was not abel to `log_model` as there is currenty no [flavor](https://mlflow.org/docs/latest/models.html#built-in-model-flavors) for **darts** (**pytorch** flavor does not work eather). But as #2092 ist still a draft, let's hope there will be a full example...

**Comment by WilliamKoch** [comment]
Looking through issues both here and in the MLFLow git, this seems an open question. I'm using MLFlow and Darts TFTModel together right now and it's partly successful picking up Lightning logs, but only saving a single _TFTModule which can't reconstitute the full Darts class.

From reading through some code, as I understand it, all of the Darts Torch Lightning based models are saved into two parts, the .pt and .pt.ckpt and it's necessary to have both to read back the weights. Is this correct?

I think the right path forward here might be to write a custom flavor (similar to how @turbotimon mentioned), which properly accounts for slight differences and supports logging, autologging, etc. and could be a MVP. Then eventually work to make sure it can support the full MLFlow API (serving inferences, docker deployments, etc.).

The code would look ~80% similar to the existing Pytorch flavor here: https://github.com/mlflow/mlflow/tree/master/mlflow/pytorch

Does this sound like the right approach?

**Comment by dennisbader** [comment]
@jakubchlapek and @daidahao are working on this 🚀
